core/include/usFrameworkEvent.h

```diff
@@ -0,0 +1,217 @@
+ * <p>
+ * <code>FrameworkEvent</code> objects are delivered to listeners connected
+ * via BundleContext::AddFrameworkListener() when an event occurs within the
+ * Framework which a user would be interested in. A <code>Type</code> code 
+ * is used to identify the event type for future extendability.
+ *
+ * @see BundleContext#AddFrameworkListener
+ */
+class US_Core_EXPORT FrameworkEvent
+{
+
+  SharedDataPointer<FrameworkEventData> d;
+
+public:
+
+  /**
+   * A type code used to identify the event type for future extendability.
+   */
+  enum class Type : unsigned int {
+
+	/**
+	 * The Framework is about to be started.
+	 * <p>
+	 * The Framework's
+	 * \link BundleActivator::Start(BundleContext*) BundleActivator Start\endlink method
+	 * is about to be called.
+	 */
+    STARTING,
+
+    /**
+     * The Framework has been started.
+     * <p>
+     * The Framework's
+     * \link BundleActivator::Start(BundleContext*) BundleActivator Start\endlink method
+     * has been executed.
+     */
+    STARTED = 1,
+
+	/**
+	 * An error has occurred.
+	 * <p>
+	 * There was an error associated with a bundle.
+	 */
+    ERROR = 2,
+
+	/**
+	 * An informational event has occurred.
+	 * <p>
+	 * There was an informational event associated with a bundle.
+	 */
+    INFO = 32,
+
+	/**
+	 * A warning has occurred.
+	 * <p>
+	 * There was a warning associated with a bundle.
+	 */
+    WARNING = 16,
+
+    /**
+     * The Framework has been stopped.
+     * <p>
+     * The Framework's
+     * \link BundleActivator::Stop(BundleContext*) BundleActivator Stop\endlink method
+     * has been executed.
+     */
+    STOPPED = 64,
+
+    /**
+     * The Framework is about to be stopped.
+     * <p>
+     * The Framework's
+     * \link BundleActivator::Stop(BundleContext*) BundleActivator Stop\endlink method
+     * is about to be called.
+     */
+    STOPPING,
+
+    /**
+     * The Framework did not stop before the wait timeout expired. 
+	 * <p>
+	 * This event is fired when the Framework did not stop before the wait timeout expired. 
+	 * The source of this event is the System Bundle.
+     */
+    WAIT_TIMEDOUT = 512
+
+  };
+
+  /**
+   * Creates an invalid instance.
+   */
+  FrameworkEvent();
+
+  ~FrameworkEvent();
+
+  /**
+   * Returns <code>false</code> if the FrameworkEvent is empty (i.e invalid) and
+   * <code>true</code> if the FrameworkEvent is not null and contains valid data.
+   */
+  bool operator!() const;
+
+  /**
+   * Creates a Framework event of the specified type.
+   *
+   * @param type The event type.
+   * @param bundle The bundle associated with the event. This bundle is also the source of the event.
+   * @param message The message associated with the event.
+   * @param exception The exception associated with this event. Should be nullptr if there is no exception.
+   */
+  FrameworkEvent(Type type, const std::shared_ptr<Bundle>& bundle, const std::string& message, const std::exception_ptr exception = nullptr);

```

Comment:
I felt that given that the OSGi spec allows an exception (i.e. ```Throwable```) object to be ```null``` and that there exists the ```INFO``` event type, which is described as informational, an exception is not always appropriate and another way to convey a message is necessary. My solution was to add the ```message``` parameter for cases where ```std::exception_ptr``` is ```nullptr```.

This is of course up for debate.

I didn't think it made sense, from a C++ developer standpoint, to provide a ```std::exception_ptr``` solely to provide an informational message.